src/ucp/core/ucp_ep.c

```diff
@@ -2791,12 +2792,29 @@ void ucp_ep_do_keepalive(ucp_ep_h ep, ucp_lane_map_t *lane_map)
         } else if (status != UCS_OK) {
             ucs_diag("unexpected return status from doing keepalive(ep=%p, "
                      "lane[%d]=%p): %s",
-                     ep, lane, ep->uct_eps[lane],
-                     ucs_status_string(status));
+                     ep, lane, ep->uct_eps[lane], ucs_status_string(status));
         }
 
-        *lane_map &= ~UCS_BIT(lane);
+        worker->keepalive.lane_map &= ~UCS_BIT(lane);
+    }
+
+    if (worker->keepalive.lane_map == 0) {
```

Comment:
is it possible that worker->keepalive.lane_map was 0 when we entered this function, so we didn't actually do any keepalive here?

Reply:
yes, it is possible, it is handled here as well

Reply:
> so we didn't actually do any keepalive here?

it's not a problem, we should skip such EPs and increment `ep_count` for them

Reply:
if we will have `begin_iter` instead of the number of EPs, we could avoid incrementing `ep_count`

Reply:
jfyi, EPs with `ep_check_map == 0` has lanes with built-in only KA (e.g. TCP)

Reply:
so maybe we should not increment the ep counter for such case?

Reply:
Then we will go through some EPs multiple times

Reply:
i mean the counter for assertion

Reply:
then we will need to add extra `if` to check whether `ep_check_map` was `0` or not, is it ok?

Reply:
yes, it's slowpath

Reply:
done